go/oasis-test-runner/scenario/e2e/gas_fees.go

```diff
@@ -0,0 +1,416 @@
+	var tee node.TEEHardware
+	err := tee.FromString(viper.GetString(cfgTEEHardware))
+	if err != nil {
+		return nil, err
+	}
+	var mrSigner *sgx.MrSigner
+	if tee == node.TEEHardwareIntelSGX {
+		mrSigner = &ias.FortanixTestMrSigner
+	}
+
+	return &oasis.NetworkFixture{
+		TEE: oasis.TEEFixture{
+			Hardware: tee,
+			MrSigner: mrSigner,
+		},
+		Network: oasis.NetworkCfg{
+			NodeBinary:          viper.GetString(cfgNodeBinary),
+			RuntimeLoaderBinary: viper.GetString(cfgRuntimeLoader),
+			EpochtimeMock:       true,
+			StakingGenesis:      "tests/fixture-data/gas-fees/staking-genesis.json",
+			LogWatcherHandlers:  DefaultBasicLogWatcherHandlers,
+		},
+		Entities: []oasis.EntityCfg{
+			oasis.EntityCfg{IsDebugTestEntity: true},
+			oasis.EntityCfg{AllowEntitySignedNodes: true},
+			oasis.EntityCfg{AllowEntitySignedNodes: true},
+			oasis.EntityCfg{AllowEntitySignedNodes: true},
+		},
+		Validators: []oasis.ValidatorFixture{
+			// Create three validators, each with its own entity so we can test
+			// if gas disbursement works correctly.
+			oasis.ValidatorFixture{Entity: 1, MinGasPrice: 1},
+			oasis.ValidatorFixture{Entity: 2, MinGasPrice: 1},
+			oasis.ValidatorFixture{Entity: 3, MinGasPrice: 1},
+		},
+	}, nil
+}
+
+func (sc *gasFeesImpl) Init(childEnv *env.Env, net *oasis.Network) error {
+	sc.net = net
+	return nil
+}
+
+func (sc *gasFeesImpl) Run(childEnv *env.Env) error {
+	if err := sc.net.Start(); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+
+	sc.logger.Info("waiting for network to come up")
+	if err := sc.net.Controller().WaitNodesRegistered(ctx, 3); err != nil {
+		return err
+	}
+	st := sc.net.Controller().Staking
+
+	// Determine initial entity balances.
+	totalEntityBalance, err := sc.getTotalEntityBalance(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Run some operations that charge fees.
+	var totalFees quantity.Quantity
+	for _, t := range []struct {
+		name string
+		fn   func(context.Context, signature.Signer, staking.Backend) (*quantity.Quantity, error)
+	}{
+		{"Transfer", sc.testTransfer},
+		{"Burn", sc.testBurn},
+		{"AddEscrow", sc.testAddEscrow},
+		{"ReclaimEscrow", sc.testReclaimEscrow},
+	} {
+		sc.logger.Info("testing operation", "op", t.name)
+
+		var fees *quantity.Quantity
+		if fees, err = t.fn(ctx, srcSigner, st); err != nil {
+			return fmt.Errorf("%s: %w", t.name, err)
+		}
+		_ = totalFees.Add(fees)
+	}
+
+	// Make sure that fees have been transferred out.
+	newTotalEntityBalance, err := sc.getTotalEntityBalance(ctx)
+	if err != nil {
+		return err
+	}
+	_ = newTotalEntityBalance.Sub(totalEntityBalance)
+
+	commonPool, err := st.CommonPool(ctx, 0)
+	if err != nil {
+		return err
+	}
+	_ = newTotalEntityBalance.Add(commonPool)
+
+	sc.logger.Info("making sure that fees have been disbursed",
+		"total_fees", totalFees,
+		"disbursed_fees", newTotalEntityBalance,
+		"common_pool", commonPool,
+	)
+
+	if newTotalEntityBalance.Cmp(&totalFees) != 0 {
```

Comment:
This should probably assert that the non-common pool disbursment is non-zero, on the off chance that something is horribly wrong and only the common pool is getting paid.

Reply:
Done.